From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Rong Bao <webmaster@csmantle.top>
Date: Wed, 14 Oct 2026 16:23:29 +0000
Subject: [PATCH 04/10] Bug 2081746 - [mips64] Implement lshift32Then{Add,Or},
 rshiftPtrThenXor, and rotateLeft64ThenXor for MIPS64.
 r=#spidermonkey-reviewers!

This is similar to https://phabricator.services.mozilla.com/D330421.

Differential Revision: https://phabricator.services.mozilla.com/D333638
---
 js/src/jit/mips64/MacroAssembler-mips64-inl.h | 31 +++++++++++++++++++
 1 file changed, 31 insertions(+)

diff --git a/js/src/jit/mips64/MacroAssembler-mips64-inl.h b/js/src/jit/mips64/MacroAssembler-mips64-inl.h
index 8828d246f3ebb241658066c20e9792109c2dfc1a..d040814fe3a0bec5b158d777fd0663336c5fc340 100644
--- a/js/src/jit/mips64/MacroAssembler-mips64-inl.h
+++ b/js/src/jit/mips64/MacroAssembler-mips64-inl.h
@@ -469,6 +469,37 @@ void MacroAssembler::rotateRight64(Register count, Register64 src,
   ma_dror(dest.reg, src.reg, count);
 }
 
+// ===============================================================
+// Shift or rotate, then combine with another register
+
+void MacroAssembler::lshift32ThenAdd(Imm32 shift, Register rhs,
+                                     Register srcDest) {
+  MOZ_ASSERT(rhs != srcDest);
+  lshift32(shift, srcDest);
+  add32(rhs, srcDest);
+}
+
+void MacroAssembler::lshift32ThenOr(Imm32 shift, Register rhs,
+                                    Register srcDest) {
+  MOZ_ASSERT(rhs != srcDest);
+  lshift32(shift, srcDest);
+  or32(rhs, srcDest);
+}
+
+void MacroAssembler::rshiftPtrThenXor(Imm32 shift, Register rhs,
+                                      Register srcDest) {
+  MOZ_ASSERT(rhs != srcDest);
+  rshiftPtr(shift, srcDest);
+  xorPtr(rhs, srcDest);
+}
+
+void MacroAssembler::rotateLeft64ThenXor(Imm32 count, Register64 rhs,
+                                         Register64 srcDest) {
+  MOZ_ASSERT(rhs != srcDest);
+  rotateLeft64(count, srcDest, srcDest, InvalidReg);
+  xor64(rhs, srcDest);
+}
+
 // ===============================================================
 // Condition functions
 
